radicle/src/cob/store.rs

```diff
@@ -187,6 +200,114 @@ where
     }
 }
 
+/// Allows operations to be batched atomically.
+#[derive(Debug)]
+pub struct Transaction<T: FromHistory> {
+    actor: ActorId,
+    start: Lamport,
+    clock: Option<Lamport>,
+    actions: Vec<T::Action>,
```

Comment:
Does making this `NonEmpty` make the rest of the code awkward to use?

Reply:
Yeah the problem is you need a different API then, eg. you can't do:
```rust
store.transaction(|tx| {
  let id = tx.foo();
  tx.bar(id);
})
```
Because it would start out empty. You also can't just pass in a list, because you want something that could be monadic here, eg. where state is threaded through so that you can refer to `OpId`s within the tx.

The more annoying thing actually is the `clock`, which also needs an "empty" state for "initial" transactions, ie. when a COB is being created. In that scenario, the first clock value is 0, so we'd have to set it to -1 as a starting value, which isn't possible, hence the `Option`.

Reply:
Right, I had a feeling as such. Makes sense :)